File: Braces.cs
using System;

namespace JobNimbus
{
    class Braces
    {
        static bool MatchingBraces(string inputString)
        {
            int openBraces = 0; //keep track of how many open braces we've encountered

            foreach (char c in inputString) //loop through each character of the given string
            {
                if (c.Equals('{'))
                {
                    openBraces++; //increment the number of open braces that need closing braces
                }

                if (c.Equals('}'))
                {
                    if (openBraces <= 0) //if we haven't encountered any open braces then we already know we have mismatched braces
                    {
                        return false;
                    }

                    openBraces--; //decrement the number of open braces that don't have a closing brace
                }
            }

            return !(openBraces > 0); //if the number of open braces is 0 return true, otherwise false
        }

        static void UnitTest(string[] testInputsArray, bool[] testOutputsArray)
        {
            Console.WriteLine("Start testing...");

            for (int i=0; i<testInputsArray.Length; i++)
            {
                string testInput = testInputsArray[i];
                bool outputActual = testOutputsArray[i];
                bool output = MatchingBraces(testInput);

                if (output != outputActual)
                {
                    Console.WriteLine($"For the string '{testInput}' the output '{output}' was given. The expected output was '{outputActual}'.");
                }
            }

            Console.WriteLine("End testing...");
        }

        static void Main(string[] args)
        {
            string[] testInputsArray = new string[] { "{}", "}{", "{{}", "", "{abc...xyz}" };
            bool[] testOutputsArray = new bool[] { true, false, false, true, true };

            UnitTest(testInputsArray, testOutputsArray);
        }
    }
}

File: braces.js
let matching_braces = function(input_string){
    let num_open_braces = 0; //keep track of how many open braces we've encountered

    for(let char of input_string){ //loop through each character of the given string
        if(char === "{"){
            num_open_braces++; //increment the number of open braces that need closing braces
        }
        
        if(char === "}"){
            if(!num_open_braces){ //if we haven't encountered any open braces then we already know we have mismatched braces
                return false;
            }

            num_open_braces--; //decrement the number of open braces that don't have a closing brace
        }
    }

    return !num_open_braces; //if the number of open braces is 0 return true, otherwise false
}

let unit_test = function(test_inputs_array, test_outputs_array){
    console.log("Start testing...");

    for(let i=0; i<test_inputs_array.length; i++){
        let test_input = test_inputs_array[i];
        let output_actual = test_outputs_array[i];
        let output = matching_braces(test_input);

        if(output != output_actual){
            console.log(`For the string '${test_input}' the output '${output}' was given. The expexted output was '${output_actual}'.`);
        }
    }

    console.log("End testing...");
}

let test_inputs_array =  ["{}", "}{", "{{}", "", "{abc...xyz}"];
let test_outputs_array = [true, false, false, true, true];
unit_test(test_inputs_array, test_outputs_array);